# $\alpha$=0.1

In [2]:
import torch
import numpy as np
import torchvision.transforms as transforms               
from torchvision.datasets import CIFAR10       
from torch.utils.data import DataLoader
import detectors
import timm
from src.temperature_scaling import ModelWithTemperature
from src.aps import aps_test

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# load fine-tuned model
model = timm.create_model("resnet34_cifar10", pretrained=True)
model.to(device)

# reprocess the images from CIFAR10
data_transform = transforms.Compose([
    transforms.ToTensor(),          # transfer to tensor
    transforms.Normalize(mean=[0.4914, 0.4822, 0.4465], std=[0.2023, 0.1994, 0.2010])  # normalize
])
# load images from CIFAR10
dataset = CIFAR10(root="../../data", train=False, download=True, transform=data_transform)

# Temperature Scaling
temp_scal_loader = DataLoader(dataset, batch_size=32, shuffle=True)
model = ModelWithTemperature(model, temperature=5.0).to(device)
model.set_temperature(temp_scal_loader)

model.eval()

# The number of times the experiment is going to be repeated
num_runs = 10

# error rate
alpha = 0.1

aps_test(model, dataset, device, num_runs=10, alpha=0.1)

Files already downloaded and verified
Before temperature - NLL: 0.331, ECE: 0.044
Optimal temperature: 4.904
After temperature - NLL: 0.882, ECE: 0.469
APS Classification, Start!

Running experiment 1/10...
q_hat = 0.4819197863340378
Total set size: 6259
Total coverage sets: 4470
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.2518
Average Coverage Rate in runs 1: 0.894

Running experiment 2/10...
q_hat = 0.4808537185192109
Total set size: 6303
Total coverage sets: 4492
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.2606
Average Coverage Rate in runs 2: 0.8984

Running experiment 3/10...
q_hat = 0.4787407666444779
Total set size: 6202
Total coverage sets: 4445
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.2404
Average Coverage Rate in runs 3: 0.889

Running experiment 4/10...
q_hat = 0.4764268398284912
Total set size: 6129
Total coverage sets: 4441
Total samples amount: 5000
Average Prediction 

# $\alpha$=0.05

In [3]:
aps_test(model, dataset, device, num_runs=10, alpha=0.05)

APS Classification, Start!

Running experiment 1/10...
q_hat = 0.5322149515151979
Total set size: 7894
Total coverage sets: 4722
Total samples amount: 5000
Average Prediction Set Size After APS in runs 1: 1.5788
Average Coverage Rate in runs 1: 0.9444

Running experiment 2/10...
q_hat = 0.5284454137086868
Total set size: 7804
Total coverage sets: 4740
Total samples amount: 5000
Average Prediction Set Size After APS in runs 2: 1.5608
Average Coverage Rate in runs 2: 0.948

Running experiment 3/10...
q_hat = 0.5294828057289125
Total set size: 7838
Total coverage sets: 4736
Total samples amount: 5000
Average Prediction Set Size After APS in runs 3: 1.5676
Average Coverage Rate in runs 3: 0.9472

Running experiment 4/10...
q_hat = 0.5287661701440812
Total set size: 7728
Total coverage sets: 4701
Total samples amount: 5000
Average Prediction Set Size After APS in runs 4: 1.5456
Average Coverage Rate in runs 4: 0.9402

Running experiment 5/10...
q_hat = 0.5365400612354279
Total set size: 805

# Result
$\alpha$=0.1
- Final Average **Prediction Set Size： 1.25**
- Final Average **Coverage: 89.43%**  

$\alpha$=0.05
- Final Average **Prediction Set Size： 1.57**
- Final Average **Coverage: 94.62%**